lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -567,177 +568,182 @@ def write(chunk)
-                GRPC::PermissionDenied,
-                # HTTP status 404 (Not Found).
-                GRPC::NotFound,
-                # HTTP status 409 (Conflict).
-                GRPC::Aborted,
-                # HTTP status 412 (Precondition Failed).
-                GRPC::FailedPrecondition,
-                # HTTP status 429 (Too Many Requests).
-                GRPC::ResourceExhausted,
-                # HTTP status 499 (Client Closed Request).
-                GRPC::Cancelled,
-                # the remaining http codes in both 4xx and 5xx category.
-                # It's debatable whether to retry or drop these log entries.
-                # This decision is made to avoid retrying forever due to client
-                # errors.
-                GRPC::Unknown
-              increment_failed_requests_count(error.code)
-              increment_dropped_entries_count(entries_count, error.code)
-              @log.warn "Dropping #{entries_count} log message(s)",
-                        error: error.to_s, error_code: error.code.to_s
-
-            else
-              # Assume it is a problem with the request itself and don't retry.
-              increment_failed_requests_count(error.code)
-              increment_dropped_entries_count(entries_count, error.code)
-              @log.error "Unknown response code #{error.code} from the" \
-                         " server, dropping #{entries_count} log message(s)",
-                         error: error.to_s, error_code: error.code.to_s
-            end
+    def write_request_function
+      @use_grpc ? 'write_request_via_grpc' : 'write_request_via_http'
+    end
 
-          # Got an unexpected error (not Google::Gax::GaxError) from the
-          # google-cloud-logging lib.
-          rescue StandardError => error
-            increment_failed_requests_count(GRPC::Core::StatusCodes::UNKNOWN)
-            increment_dropped_entries_count(entries_count,
-                                            GRPC::Core::StatusCodes::UNKNOWN)
-            @log.error "Unexpected error type #{error.class.name} from the" \
-                       " client library, dropping #{entries_count}" \
-                       ' log message(s)',
-                       error: error.to_s
-          end
-        else
-          begin
-            write_request =
-              Google::Apis::LoggingV2::WriteLogEntriesRequest.new(
-                log_name: log_name,
-                resource: group_level_resource,
-                labels: group_level_common_labels,
-                partial_success: @partial_success,
-                entries: entries)
-            entries_count = entries.length
-            client.write_entry_log_entries(
-              write_request,
-              options: { api_format_version: '2' }
-            )
+    def write_request_via_grpc(client, request)
+      entries_count = request.entries.length
+      labels_utf8_pairs = request.labels.map do |k, v|
```

Comment:
No longer "pairs", right? This is now a hash. How about `labels_utf8` or `utf8_encoded_labels`?

Reply:
Nice catch. Rename `labels_utf8_pairs` to `utf8_encoded_labels`.